[quant][pt2e] Support quantizer API in prepare_pt2e_quantizer

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97994

Summary:
This PR added a quantizer API to prepare_pt2e_quantizer, which enables user to annotate the nodes in the graph
directly to configure quantization, instead of relying on QConfigMapping, please see test cases in
test_quantize_pt2e.py for examples. Also added a prototype for QNNPackQuantizer, that will be modified later
to fully support different quantization capabilities of QNNPack/XNNPack

The goal for introducing quantizer is to add flexibility to the quantization API to allow modeling users and backend developers to express their quantization intentions programmably, which will free architecture optimization team from supporting different use cases in the core API in the future, as a concrete example, we used to have https://pytorch.org/docs/master/generated/torch.ao.quantization.qconfig_mapping.QConfigMapping.html#torch.ao.quantization.qconfig_mapping.QConfigMapping as the API for users to express their intent for quantization in fx graph mode quantization, and it has some fancy options like `set_module_name_regex` and `set_module_name_object_type_order`, this is not needed for all backends and adds burden of maintenance to AO team, in the quantizer API we will move these APIs to a backend specific `Quantizer` that needs this feature, and all the backends or even advanced modeling users can implement their own quantizer to express their intent for quantization through annotating the nodes, for example, to express the quantization intention of quantizing a convolution node, a user will find the convolution node in the graph and do:
```
operator_spec = qnnpack_quantizer.get_default_per_channel_symmetric_qnnpack_operator_spec()
conv_node.meta["target_dtype_info"] = {
    "input_act_obs_or_fq_ctr": _get_act_obs_or_fq_ctr(operator_spec),
    "weight_obs_or_fq_ctr": _get_weight_obs_or_fq_ctr(operator_spec)
    "bias_obs_or_fq_ctr": _get_bias_obs_or_fq_ctr(operator_spec),
    "output_act_obs_or_fq_ctr": _get_act_obs_or_fq_ctr(operator_spec),
    # TODO: validation of weight_index must be set if weight_obs_or_fq_ctr is set
    "weight_index": 1,
    # TODO: validation of bias_index must be set if bias_obs_or_fq_ctr is set
    "bias_index": 2,
}
```
each backend will introduce their own quantizer, e.g. QNNPackQuantizer, which may expose more convenient APIs for modeling users to configure the annotation, and different quantizer can compose with each other to annotate the graph correctly for quantization.

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_simple_quantizer
python test/test_quantization.py TestQuantizePT2E.test_qnnpack_quantizer_conv

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
[quant][pt2e] Add quantizer API to prepare_pt2e

Summary:
This PR added a quantizer API to prepare_pt2e, which enables user to annotate the nodes in the graph
directly to configure quantization, instead of relying on QConfigMapping, please see test cases in
test_quantize_pt2e.py for examples.

Test Plan:
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented on torch/ao/quantization/_pt2e/_quantizer.py:
```diff
@@ -0,0 +1,18 @@
+
+class Quantizer(ABC):
+
+    def __init__(self, backend_config: Optional[BackendConfig] = None):
```
we should not have backend config as user facing api at all

jerryzh168 replied:
sure, what is the recommendation? are you talking about removing this from the argument and just have backend assign the member variable directly?

supriyar commented:
> please see test cases in test_quantize_pt2e.py for examples

the test file doesn't seem to be included in this PR?

jerryzh168 commented:
> > please see test cases in test_quantize_pt2e.py for examples
> 
> the test file doesn't seem to be included in this PR?

yeah this PR is WIP, just put up earlier so people know what it will look like

jerryzh168 committed:
Update on "[quant][pt2e] Add quantizer API to prepare_pt2e"


Summary:
This PR added a quantizer API to prepare_pt2e, which enables user to annotate the nodes in the graph
directly to configure quantization, instead of relying on QConfigMapping, please see test cases in
test_quantize_pt2e.py for examples.

Test Plan:
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented on torch/ao/quantization/quantizer/qnnpack_quantizer.py:
```diff
@@ -0,0 +1,145 @@
+
+
+@dataclass
+class TensorSpec:
```
Why is this called tensorSpec. It should be QuantizationSpec.

jerryzh168 replied:
this is initially called QuantizationParam, and OperatorSpec is called QuantizationSpec, feels a bit confusing though, but this file is supposed to be a starting point/prototype, feel free to modify since right now qnnpack/xnnpack is fully owning this

jerryzh168 replied:
can you postpone moving to base quantizer a bit, I would like to discuss with other people and gather more user data from backend developers before deciding

kimishpatel replied:
having annotate and validate as the only thing in base quantizer makes no sense to me. We at least have to enforce implementation of quantizer to expose, what quantization spec and modules/op they support.

kimishpatel replied:
>  would like to discuss with other people and gather more user data from backend developers before deciding

FOr this can you concretely identify what the concerns are?

jerryzh168 replied:
clarified in chat, I'm fine with either as long as we are open to change according to data in the future

kimishpatel commented on torch/ao/quantization/quantizer/qnnpack_quantizer.py:
```diff
@@ -0,0 +1,145 @@
+
+
+def supported_symmetric_quantized_operators():
+    supported_module = [torch.nn.Conv2d, torch.nn.functional.conv2d]
```
I would like to say this: quant supports nn.Conv2d: This means it supports nn.Conv2d but also funcitonal.conv2d. This is simpler to understand IMO and hides complexity. 

You can certainly argue that it is a bit confusing because when API says nn.Conv2d is supported, it is not saying it supports functional.conv2d, but all the modules and functional variants eventually map to functional variants. So I would like to keep it this way and maybe get more input to say if we need to enumerate functional variants here as well.

kimishpatel commented on torch/ao/quantization/quantizer/qnnpack_quantizer.py:
```diff
@@ -0,0 +1,145 @@
+)
+
+
+def supported_symmetric_quantized_operators():
```
I dont know why are these free methods. THis is exactly why I wanted them to be on Quantizer. I think its fine for this prototype but I am gonna add those to quantizer.

jerryzh168 replied:
I think we can discuss, I'd like to start with qnnpack for now, we can decide before we advertise this to other backends if we want to put those in quantizer or not

kimishpatel commented on torch/ao/quantization/quantizer/qnnpack_quantizer.py:
```diff
@@ -0,0 +1,145 @@
+            raise ValueError("Ch_axis is < 0.")
+
+
+OperatorSpec = typing.NamedTuple(
```
Why OperatorSpec? I think QuantizationSpec is also good. Although given that this in quantization codebase, it does not have to be have Quantization in the name, but not sure what OperatorSpec means for user. For implementation it makes sense because you are talking about aten op.

jerryzh168 replied:
yeah this is just something as a starting point, please feel free to modify to something makes more sense to qnnpack users, I think the current naming mainly reflects the granularity of the spec, one is for tensor, the other is for operator

kimishpatel replied:
I dont think we should distinguish between qnnpack users or other backends. I dont want to block this PR but all these have to be in base quantizer

kimishpatel commented on torch/ao/quantization/quantizer/qnnpack_quantizer.py:
```diff
@@ -0,0 +1,145 @@
+        return self
+
+    def set_operator_type(self, obct_type: str, operator_spec: OperatorSpec) -> OperatorSpecConfig:
+        self.operator_type_specs[operator_type] = operator_spec
```
What is `operator_type`? Did you mean `obct_type`?

kimishpatel commented on torch/ao/quantization/_pt2e/prepare.py:
```diff
@@ -0,0 +1,115 @@
+        if node.op == "placeholder":
+            pass
+        elif node.op in ("call_module", "call_method", "call_function"):
+            this_node_dtype_info = node.meta["target_dtype_info"]
```
We dont have to do it now but "target_dtype_info" does not seem to be related to quantization. Maybe just called it "quantization_spec"

jerryzh168 replied:
yeah, we can change the name a bit later, I'll add a TODO

kimishpatel commented on torch/ao/quantization/_pt2e/prepare.py:
```diff
@@ -0,0 +1,115 @@
+            else:
+                output_is_a_tensor = this_node_dtype_info is not None
+
+            skip_inserting_observers = (
```
Dont you want to check if the input_activation is a valid obsever?

jerryzh168 replied:
you mean not None? it's probably handled in `_maybe_insert_input_observers_for_node` already

kimishpatel replied:
No, I mean you wont call `_maybe_insert_input_observers_for_node` if `skip_inserting_observers`. But `skip_inserting_observers` only accounts for output_is_a_tensor.

jerryzh168 replied:
oh I see, this logic is copied from fx graph mode quant code base, this is probably used to handle ops like "tensor.size()" etc. I'm not exactly sure what should be the rule for input activations but looks like it's not needed right now

kimishpatel commented on torch/ao/quantization/_pt2e/prepare.py:
```diff
@@ -0,0 +1,115 @@
+                node,
+                None, # qconfig
+                model,
+                named_modules,
```
What is the use of `named_module`?

jerryzh168 replied:
this is needed to store a mapping from observer name to observer instance right now: https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fx/prepare.py#L663 maybe it can be improved later after refactoring

kimishpatel commented on torch/ao/quantization/_pt2e/prepare.py:
```diff
@@ -0,0 +1,115 @@
+                model.graph,
+                None, # qhandler
+                PrepareCustomConfig(),
+                None, # backend_config
```
How would this work? I thought if you reuse `_maybe_insert_input_observers_for_node`, you will need a backend_config?

jerryzh168 replied:
yeah it turns out the only remaining thing we read from backend_config now is just the index of weight and bias, this PR removes it and puts them in the annotation dictionary instead: https://github.com/pytorch/pytorch/pull/98094

kimishpatel commented on torch/ao/quantization/_pt2e/prepare.py:
```diff
@@ -0,0 +1,115 @@
+            #
+            # We need to save orig users before updating uses because
+            # the list of users will change as we update uses
+            orig_users = list(node.users.keys())
+            for user_node in orig_users:
+                if user_node is maybe_output_obs_node:
+                    continue
+                user_node.replace_input_with(node, maybe_output_obs_node)
+            _is_observer_in_same_graph_ = _is_observer_in_same_graph(node, named_modules)
```
Shouldnt all of this happen as part of `_maybe_insert_output_observer_for_node`? If not at least factor it out in a different function where you call `_maybe_insert_output_observer_for_node` and also update uses?

jerryzh168 replied:
makes sense, can I do this in a separate PR, I think we should be able to just move these things to `_maybe_insert_output_observer_for_node` and update all callsites

kimishpatel commented on torch/ao/quantization/_pt2e/prepare.py:
```diff
@@ -0,0 +1,115 @@
+            # for general tensor value ops, we modify the graph
+            # to make all inputs and outputs use the first input's
+            # observer
+            if (is_general_tensor_value_op and _is_observer_in_same_graph_) or \
```
`is_general_tensor_value_op` not defined.

jerryzh168 replied:
yeah will update

kimishpatel commented on torch/ao/quantization/_pt2e/prepare.py:
```diff
@@ -0,0 +1,115 @@
+
+    model = GraphModule(model, model.graph)
+
+    _save_state(
```
Why save_state?

jerryzh168 replied:
convert code is expecting this, a lot of things are actually not needed for pt2e flow, but we can clean up later I think

jerryzh168 replied:
fx graph mode quantization has more features than pt2e right now, e.g. qat support, standalone_module, custom_module, it is also working with modules, a lot of states are for supporting these features

kimishpatel reviewed (changes requested):
Left a few comments. Also move quantizer to _pt2 folder?

jerryzh168 commented on torch/ao/quantization/_pt2e/prepare.py:
```diff
@@ -0,0 +1,115 @@
+                user_node.replace_input_with(node, maybe_output_obs_node)
+            _is_observer_in_same_graph_ = _is_observer_in_same_graph(node, named_modules)
+
+            # for general tensor value ops, we modify the graph
+            # to make all inputs and outputs use the first input's
+            # observer
```
probably don't need this, will remove

jerryzh168 commented on torch/ao/quantization/_pt2e/prepare.py:
```diff
@@ -0,0 +1,115 @@
+    _save_state(
+        model,
+        {},  # node_name_to_qconfig
+        node_name_to_scope,
```
similar to other states, convert code expects this right now, but I think we may be able to read from node.meta now if we have "nn_module_stack" or something stored there

jerryzh168 committed:
Update on "[quant][pt2e] Add quantizer API to prepare_pt2e"


Summary:
This PR added a quantizer API to prepare_pt2e, which enables user to annotate the nodes in the graph
directly to configure quantization, instead of relying on QConfigMapping, please see test cases in
test_quantize_pt2e.py for examples.

Test Plan:
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Add quantizer API to prepare_pt2e"


Summary:
This PR added a quantizer API to prepare_pt2e, which enables user to annotate the nodes in the graph
directly to configure quantization, instead of relying on QConfigMapping, please see test cases in
test_quantize_pt2e.py for examples.

Test Plan:
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Add quantizer API to prepare_pt2e"


Summary:
This PR added a quantizer API to prepare_pt2e, which enables user to annotate the nodes in the graph
directly to configure quantization, instead of relying on QConfigMapping, please see test cases in
test_quantize_pt2e.py for examples.

Test Plan:
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Support quantizer API in prepare_pt2e_quantizer"


Summary:
This PR added a quantizer API to prepare_pt2e_quantizer, which enables user to annotate the nodes in the graph
directly to configure quantization, instead of relying on QConfigMapping, please see test cases in
test_quantize_pt2e.py for examples. Also added a prototype for QNNPackQuantizer, that will be modified later
to fully support different quantization capabilities of QNNPack/XNNPack

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_simple_quantizer
python test/test_quantization.py TestQuantizePT2E.test_qnnpack_quantizer_conv

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -593,7 +593,7 @@ def _maybe_insert_input_observer_for_arg_or_kwarg(
         is_weight = node_arg_is_weight(node, arg)
 
-        reuse_input_obs_or_fq = node.meta["target_dtype_info"]["reuse_input_obs_or_fq"]
+        reuse_input_obs_or_fq = node.meta["target_dtype_info"].get("reuse_input_obs_or_fq", False)
```
can we spell out why the production code path is changed?

I think for pt2e APIs, stamping to unblock is reasonable.  For changes in reusable code, it's good to minimize them as well as spell out what they are.

jerryzh168 replied:
sure, I can add a comment, this is for pt2e path since not all nodes is using this key right now

jerryzh168 committed:
Update on "[quant][pt2e] Support quantizer API in prepare_pt2e_quantizer"


Summary:
This PR added a quantizer API to prepare_pt2e_quantizer, which enables user to annotate the nodes in the graph
directly to configure quantization, instead of relying on QConfigMapping, please see test cases in
test_quantize_pt2e.py for examples. Also added a prototype for QNNPackQuantizer, that will be modified later
to fully support different quantization capabilities of QNNPack/XNNPack

The goal for introducing quantizer is to add flexibility to the quantization API to allow modeling users and backend developers to express their quantization intentions programmably, which will free architecture optimization team from supporting different use cases in the core API in the future, as a concrete example, we used to have https://pytorch.org/docs/master/generated/torch.ao.quantization.qconfig_mapping.QConfigMapping.html#torch.ao.quantization.qconfig_mapping.QConfigMapping as the API for users to express their intent for quantization in fx graph mode quantization, and it has some fancy options like `set_module_name_regex` and `set_module_name_object_type_order`, this is not needed for all backends and adds burden of maintenance to AO team, in the quantizer API we will move these APIs to a backend specific `Quantizer` that needs this feature, and all the backends or even advanced modeling users can implement their own quantizer to express their intent for quantization through annotating the nodes, for example, to express the quantization intention of quantizing a convolution node, a user will find the convolution node in the graph and do:
```
operator_spec = qnnpack_quantizer.get_default_per_channel_symmetric_qnnpack_operator_spec()
conv_node.meta["target_dtype_info"] = {
    "input_act_obs_or_fq_ctr": _get_act_obs_or_fq_ctr(operator_spec),
    "weight_obs_or_fq_ctr": _get_weight_obs_or_fq_ctr(operator_spec)
    "bias_obs_or_fq_ctr": _get_bias_obs_or_fq_ctr(operator_spec),
    "output_act_obs_or_fq_ctr": _get_act_obs_or_fq_ctr(operator_spec),
    # TODO: validation of weight_index must be set if weight_obs_or_fq_ctr is set
    "weight_index": 1,
    # TODO: validation of bias_index must be set if bias_obs_or_fq_ctr is set
    "bias_index": 2,
}
```
each backend will introduce their own quantizer, e.g. QNNPackQuantizer, which may expose more convenient APIs for modeling users to configure the annotation, and different quantizer can compose with each other to annotate the graph correctly for quantization.

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_simple_quantizer
python test/test_quantization.py TestQuantizePT2E.test_qnnpack_quantizer_conv

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Support quantizer API in prepare_pt2e_quantizer"


Summary:
This PR added a quantizer API to prepare_pt2e_quantizer, which enables user to annotate the nodes in the graph
directly to configure quantization, instead of relying on QConfigMapping, please see test cases in
test_quantize_pt2e.py for examples. Also added a prototype for QNNPackQuantizer, that will be modified later
to fully support different quantization capabilities of QNNPack/XNNPack

The goal for introducing quantizer is to add flexibility to the quantization API to allow modeling users and backend developers to express their quantization intentions programmably, which will free architecture optimization team from supporting different use cases in the core API in the future, as a concrete example, we used to have https://pytorch.org/docs/master/generated/torch.ao.quantization.qconfig_mapping.QConfigMapping.html#torch.ao.quantization.qconfig_mapping.QConfigMapping as the API for users to express their intent for quantization in fx graph mode quantization, and it has some fancy options like `set_module_name_regex` and `set_module_name_object_type_order`, this is not needed for all backends and adds burden of maintenance to AO team, in the quantizer API we will move these APIs to a backend specific `Quantizer` that needs this feature, and all the backends or even advanced modeling users can implement their own quantizer to express their intent for quantization through annotating the nodes, for example, to express the quantization intention of quantizing a convolution node, a user will find the convolution node in the graph and do:
```
operator_spec = qnnpack_quantizer.get_default_per_channel_symmetric_qnnpack_operator_spec()
conv_node.meta["target_dtype_info"] = {
    "input_act_obs_or_fq_ctr": _get_act_obs_or_fq_ctr(operator_spec),
    "weight_obs_or_fq_ctr": _get_weight_obs_or_fq_ctr(operator_spec)
    "bias_obs_or_fq_ctr": _get_bias_obs_or_fq_ctr(operator_spec),
    "output_act_obs_or_fq_ctr": _get_act_obs_or_fq_ctr(operator_spec),
    # TODO: validation of weight_index must be set if weight_obs_or_fq_ctr is set
    "weight_index": 1,
    # TODO: validation of bias_index must be set if bias_obs_or_fq_ctr is set
    "bias_index": 2,
}
```
each backend will introduce their own quantizer, e.g. QNNPackQuantizer, which may expose more convenient APIs for modeling users to configure the annotation, and different quantizer can compose with each other to annotate the graph correctly for quantization.

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_simple_quantizer
python test/test_quantization.py TestQuantizePT2E.test_qnnpack_quantizer_conv

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge

jerryzh168 commented:
@pytorchbot merge

kimishpatel commented on torch/ao/quantization/_quantize_pt2e.py:
```diff
@@ -45,6 +47,38 @@ def prepare_pt2e(
+    # TODO: check qconfig_mapping to make sure conv and bn are both configured
+    # to be quantized before fusion
+    # TODO: (maybe) rewrite this with subgraph_rewriter
+    _fuse_conv_bn_(model)
```
THis is only valid for ptq?

andrewor14 replied:
Yes

kimishpatel commented on torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py:
```diff
@@ -0,0 +1,266 @@
+]
+
+# TODO: maybe remove torch.float32
+SUPPORTED_DTYPES = [torch.uint8, torch.int8, torch.int32, torch.float16, torch.float32]
```
nit: float16 is not something natively supported in executorch

kimishpatel commented on torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py:
```diff
@@ -0,0 +1,266 @@
+            "output_act_obs_or_fq_ctr": PlaceholderObserver.with_args(dtype=torch.float),
+        }
+        for node in model.graph.nodes:
+            node.meta["target_dtype_info"] = copy.deepcopy(_DEFAULT_TARGET_DTYPE_INFO)
```
Why do we need observer for every single node?

kimishpatel commented on torch/ao/quantization/_pt2e/prepare.py:
```diff
@@ -0,0 +1,123 @@
+
+    named_modules = dict(model.named_modules(remove_duplicate=False))
+
+    _maybe_insert_input_observers_for_node(
```
If observer needs to be inserted for input, does this also do the appropriate graph rewrite?

kimishpatel commented on test/quantization/fx/test_quantize_pt2e.py:
```diff
@@ -133,6 +135,109 @@ def forward(self, x):
+            aten_graph=True,
+            tracing_mode="real",
+        )
+        m = prepare_pt2e_quantizer(m, BackendAQuantizer())
```
we should test for observer insertion as well?
